src/storage/servers.js

```diff
@@ -0,0 +1,96 @@
+const db = require('../storage/DataAccess');
+const { connectToServer } = require('../utils/client');
+
+const SERVERS = 'servers';
+
+/**
+ * Fields for Servers Data Types:
+ *
+ * Server:
+ *  @param {string} id - the id of the server
+ *  @param {string} name - human readable name
+ *  @param {string} endpoint: - FHIR base url
+ *  @param {string} type - 'KA' | 'EHR' | 'PHA' - type of server the endpoint is
+ *  @param {string} clientId - the client ID assigned to this app during registration
+ *  @param {string} token - the last access token received from authorization
+ *  @param {number} tokenExp - the time the token expires (UTC)
+ *  @param {string} key - the last key
+ *
+ */
+
+function addServer(server) {
+  db.upsert(SERVERS, server, s => s.id === server.id);
+}
+
+function getServers() {
+  return db.select(SERVERS, () => true);
+}
+
+function getServerById(id) {
+  return db.select(SERVERS, s => s.id === id)[0];
+}
+
+function getServerByUrl(url) {
+  return db.select(SERVERS, s => s.endpoint === url)[0];
+}
+
+function deleteServer(id) {
+  db.delete(SERVERS, s => s.id === id);
+}
+
+function addClientId(server, clientId) {
+  server.clientId = clientId;
+  addServer(server);
+}
+
+function getClientId(server) {
+  return this.getServerById(server.id).clientId;
+}
+
+function addAccessToken(server, token, tokenExp) {
+  server.token = token;
+  server.tokenExp = tokenExp;
+  addServer(server);
+}
+
+async function getAccessToken(server) {
+  const result = this.getServerById(server.id);
+  if (!result.token || result.tokenExp < Date.now()) {
+    // create a new token if possible
+    try {
+      const token = await connectToServer(server.endpoint);
+      // expires_in is time until expiration in seconds, Date.now() is in milliseconds
+      const exp = Date.now() + token.expires_in * 1000;
+      db.addAccessToken(server, token.access_token, exp);
+      return token;
+    } catch (e) {
+      clearAccessToken(server);
+      console.error(e);
+      return null;
```

Comment:
So I didn't realize this at first, but after seeing the error in the console from the unit test, I'm wondering if catching the exception and returning null is the right approach here. This is the only place where we call out to the server to get the new token right? So if this returns null that's an exceptional situation, and there's nothing we can do to fix that. I feel like that should stay an exception, still clear the token though, and we should have a handler somewhere higher up that logs it (that handler part would be a separate task).